openedx/core/djangoapps/course_roles/helpers.py

```diff
@@ -0,0 +1,189 @@
+    """
+    Check if a user has all of the given permissions in a course.
+    """
+    if not use_permission_checks():
+        return False
+    return all(user_has_permission_course(user, permission, course_key) for permission in permissions)
+
+
+@request_cached()
+def user_has_permission_list_any_course(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        course_key: CourseKey
+):
+    """
+    Check if a user has ANY of the given permissions in a course.
+    """
+    return any(user_has_permission_course(user, permission, course_key) for permission in permissions)
+
+
+@request_cached()
+def user_has_permission_org(
+        user: Union[User, AnonymousUser],
+        permission: Union[CourseRolesPermission, str],
+        organization_name: str
+):
+    """
+    Check if a user has a permission for all courses in an organization.
+    """
+    if not use_permission_checks():
+        return False
+    elif isinstance(user, AnonymousUser):
+        return False
+    if isinstance(permission, CourseRolesPermission):
+        permission = permission.value.name
+    return UserRole.objects.filter(
+        user=user,
+        role__permissions__name=permission,
+        course__isnull=True,
+        org__name=organization_name,
+    ).exists()
+
+
+@request_cached()
+def user_has_permission_list_org(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        organization_name: str
+):
+    """
+    Check if a user has ALL of the given permissions for all courses in an organization.
+    """
+    if not use_permission_checks():
+        return False
+    return all(
+        user_has_permission_org(user, permission, organization_name) for permission in permissions
+    )
+
+
+@request_cached()
+def user_has_permission_course_org(
+        user: Union[User, AnonymousUser],
+        permission: Union[CourseRolesPermission, str],
+        course_key: CourseKey
+):
+    """
+    Check if a user has a permission for all courses in an organization or for a specific course.
+    """
+    if not use_permission_checks():
+        return False
+    elif isinstance(user, AnonymousUser):
+        return False
+    if isinstance(permission, CourseRolesPermission):
+        permission = permission.value.name
+    return (user_has_permission_course(user, permission, course_key) or
+            user_has_permission_org(user, permission, course_key.org)
+            )
+
+
+@request_cached()
+def user_has_permission_list_course_org(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        course_key: CourseKey
+):
+    """
+    Check if a user has all of the given permissions for all courses in an organization or for a specific course.
+    """
+    if not use_permission_checks():
+        return False
+    return all(
+        user_has_permission_course_org(user, permission, course_key)
```

Comment:
For performance reasons, when you are defining an "access one thing" vs. "access many things" calls, you generally want to either have their implementations be separate, or define the "access one thing" as a special case of "access many" rather than have "access many" call "access one" in serial. It will result in fewer queries that way. For example, if there are ten permissions to be checked here, this code going to make ten database calls.

This is a sketch of code to show what I mean (I only sketched it out to illustrate the idea–I haven't tried to test it, so it probably has bugs):

```python
def user_has_permission(
    user: User | AnonymousUser,
    course_key: CourseKey,
    permission: CourseRolesPermission,
) -> bool:
    """
    Does the user have a permission for a given Course?

    This will check for permissions that the user might have because of their
    course-specific, org, or site level roles. It is semantically equivalent to
    calling user_has_all_permissions with only one permission to be checked. We
    just have it as a convenience function because checking against a single
    permission is so common.
    """
    return user_has_all_permissions(user, course_key, {permission})

def user_has_all_permissions(
    user: User | AnonymousUser,
    course_key: CourseKey,
    permissions: Iterable[CourseRolesPermission],
) -> bool:
    """
    Does the user have all the requested permissions for a given Course?

    Accepts any Iterable of CourseRolesPermission. This will check for
    permissions that the user might have because of their course-specific, org,
    or site level roles.
    """
    if not use_permission_checks():
        return False

    if not isinstance(permissions, set):
        permissions = set(permissions)

    return permissions <= all_user_permissions_for_course(user, course_key)

def user_has_any_permissions(
    user: User | AnonymousUser,
    course_key: CourseKey,
    permissions: Iterable[CourseRolesPermission],
) -> bool:
    """
    Does the user have any of the requested permissions for a given Course?

    Accepts any Iterable of CourseRolesPermission. This will check for
    permissions that the user might have because of their course-specific, org,
    or site level roles.
    """
    if not use_permission_checks():
        return False

    if not isinstance(permissions, set):
        permissions = set(permissions)

    return bool(permissions & all_user_permissions_for_course(user, course_key))

def all_user_permissions_for_course(
    user: User | AnonymousUser,
    course_key: CourseKey
) -> set[CourseRolesPermission]:
    """
    Get all of a user's permissions for a course.

    This includes any permissions a user has because of their org or instance
    level roles.

    Using a RequestCache manually instead of via the decorator because I'm not
    sure how the decorator will handle AnonymousUser serialization.
    """
    cache = RequestCache("course_roles")
    cache_key = f"all_user_permissions_for_course:{user.id}:{course_key}"
    is_found, cached_value = cache.get_cached_response(cache_key)
    if is_found:
        return cached_value

    if not CourseOverview.course_exists(course_key):
        raise ValueError('course does not exist')

    if isinstance(user, AnonymousUser):
        user_query = Q(user__isnull=True)
    elif isinstance(user, User):
        user_query = Q(user__id=user.id)

    permissions_qset = UserRole.objects.filter(
        user_query,
        (
            # Course-specific roles
            Q(course=course_key) |
            # Org-wide roles that apply to this course
            (Q(course__isnull=True) & Q(org__name=course_key.org)) |
            # Instance-wide roles
            Q(org__isnull=True)
        )
    )

    # The actual code for this needs to check to see if the permission_name
    # really exists in the Enum and to gracefully log an error message and
    # continue on if it does. When a new permission is added, there is going
    # to be a brief period of time where a new permission exists in the database
    # that older web workers won't recognize because they're still running old
    # code.
    permissions = set(
        # This might not work because of casing. Can we make permission_name
        # upper case like the enum names are?
        CourseRolesPermission[permission_name]
        for permission_name
        in permissions_qset.values_list('role__permissions__name', flat=True)
    )
    cache.set(cache_key, permissions)

    return permissions

```

Reply:
In the above code scenario, only the `all_user_permissions_for_course` function needs to be cached, because it's the only one making the database query. The set operations to compare the requested permissions vs. the cached results is very fast, and doesn't need caching. This also gives a convenient place to stub permissions into tests if dealing with the cache becomes annoying, because you can mock this one function and return whatever list of permissions you want for this user for a given test, and all the functions that rely on it can use that value without touching the database or cache.